src/stackwalk.c

```diff
@@ -1166,10 +1166,20 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
 }
 
 extern int gc_first_tid;
+extern int jl_inside_heartbeat_thread(void);
+extern int jl_heartbeat_pause(void);
+extern int jl_heartbeat_resume(void);
 
-// Print backtraces for all live tasks, for all threads, to jl_safe_printf stderr
+// Print backtraces for all live tasks, for all threads, to jl_safe_printf
+// stderr. This can take a _long_ time!
 JL_DLLEXPORT void jl_print_task_backtraces(int show_done) JL_NOTSAFEPOINT
 {
+    // disable heartbeats to prevent heartbeat loss while running this,
+    // unless this is called from the heartbeat thread
+    if (!jl_inside_heartbeat_thread()) {
+        jl_heartbeat_pause();
+    }
```

Comment:
🤔 Why the if-check?

If a heartbeat loss causes us to print backtraces, could the backtraces we're printing cause another heartbeat loss?

I think I remember you said the answer is "no" when we discussed over zoom, but it would be nice to clarify that in the comments, here.

Reply:
I thought I'd written that up. Sorry, I'll add a comment. If this is called from the heartbeat thread, there will be no additional heartbeat loss because it is the heartbeat thread itself that determines whether there is heartbeat loss. So when it returns from here, the counter for heartbeats missed is unchanged.